src/metabase/pulse/markdown.clj

```diff
@@ -0,0 +1,320 @@
+                                content))
+        joined-content   (str/join resolved-content)]
+    (case tag
+      :document
+      joined-content
+
+      :paragraph
+      (str joined-content "\n")
+
+      :soft-line-break
+      " "
+
+      :hard-line-break
+      "\n"
+
+      (:heading :bold)
+      (str "*" joined-content "*")
+
+      :italic
+      (str "_" joined-content "_")
+
+      :code
+      (str "`" joined-content "`")
+
+      :codeblock
+      (str "```\n" joined-content "```")
+
+      :blockquote
+      (let [lines (str/split-lines joined-content)]
+        (str/join "\n" (map #(str ">" %) lines)))
+
+      :link
+      (let [resolved-uri (resolve-uri (:href attrs))]
+        (if (contains? #{:image :image-ref} (:tag (first content)))
+          ;; If this is a linked image, add link target on separate line after image placeholder
+          (str joined-content "\n(" resolved-uri ")")
+          (str "<" resolved-uri "|" joined-content ">")))
+
+      :link-ref
+      (if-let [resolved-uri (resolve-uri (-> attrs :reference :attrs :url))]
+        (str "<" resolved-uri "|" joined-content ">")
+        joined-content)
+
+      :auto-link
+      (str "<" (:href attrs) ">")
+
+      :mail-link
+      (str "<" (:address attrs) ">")
+
+      ;; list items might have nested lists or other elements, which should have their indentation level increased
+      (:unordered-list-item :ordered-list-item)
+      (let [indented-content (->> (rest resolved-content)
+                                  str/join
+                                  str/split-lines
+                                  (map #(str "    " %))
+                                  (str/join "\n"))]
+        (if-not (str/blank? indented-content)
+          (str (first resolved-content) indented-content "\n")
+          joined-content))
+
+      :unordered-list
+      (str/join (map #(str "• " %) resolved-content))
+
+      :ordered-list
+      (str/join (map-indexed #(str (inc %1) ". " %2) resolved-content))
+
+      ;; Replace images with text that links to source, including alt text if available
+      :image
+      (let [{:keys [src alt]} attrs]
+        (if (str/blank? alt)
+          (str "<" src "|[Image]>")
+          (str "<" src "|[Image: " alt "]>")))
+
+      :image-ref
+      (let [src (-> attrs :reference :attrs :url)
+            alt joined-content]
+        (if (str/blank? alt)
+          (str "<" src "|[Image]>")
+          (str "<" src "|[Image: " alt "]>")))
+
+      :html-entity
+      (some->> content
+               (get @html-entities)
+               (:characters))
+
+      joined-content)))
+
+(def ^:private ^:dynamic *slack-mrkdwn-length-limit*
+  3000)
+
+(defn- truncate-mrkdwn
+  "If a string is greater than Slack's length limit, truncates it to fit the limit and
+  adds an ellipsis character to the end."
+  [mrkdwn]
+  (if (> (count mrkdwn) *slack-mrkdwn-length-limit*)
+    (-> mrkdwn
+        (subs 0 (dec *slack-mrkdwn-length-limit*))
```

Comment:
If the limit is 3000, is there an issue with taking 2999 characters and then adding three characters to end up at 3002? Should this ensure the full ellipses is included at the 3000 character limit? Could be misleading if we then add a single period rather than the three to indicate truncation and continuation.

Reply:
Also, I think this gets at what i was getting at with `:slack` vs `:mrkdwn`. I think we should have this namespace pure, and no notion of slack and limits come in. Just a pure namespace that knows how to parse the markdown language and how to translate to the mrkdwn language. The slack pulse/render code can be the one that has to know about vulgar concepts like truncating overly long messages, but our parsing code can be kept blisfully unaware of these things. It simple translates one language into another.

Reply:
Fortunately, the ellipsis is a single character: `…`

Sure, I'll move this to `metabase.pulse`.

Reply:
oh right. good answer on that being a single character :)